### ISSUE TITLE
WARNING: JNI local refs: 35, exceeds capacity: 34

### ISSUE THREAD

**Issue by mknejp** [dropbox/djinni#153]
After upgrading JDK 8 I get these two warnings when running the test-suite:
```
     [java] WARNING: JNI local refs: 35, exceeds capacity: 34
     [java] 	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
     [java] 	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1938)
     [java] 	- locked <0x000000076ab067d0> (a java.util.Vector)
     [java] 	- locked <0x000000076ab55718> (a java.util.Vector)
     [java] 	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1821)
     [java] 	at java.lang.Runtime.load0(Runtime.java:809)
     [java] 	- locked <0x000000076ab1ae20> (a java.lang.Runtime)
     [java] 	at java.lang.System.load(System.java:1086)
     [java] 	at com.dropbox.djinni.NativeLibLoader.loadLibrary(NativeLibLoader.java:171)
     [java] 	at com.dropbox.djinni.NativeLibLoader.loadLibsFromLocalPath(NativeLibLoader.java:95)
     [java] 	at com.dropbox.djinni.NativeLibLoader.loadLibs(NativeLibLoader.java:80)
     [java] 	at com.dropbox.djinni.test.AllTests.main(AllTests.java:28)
     [java] Oct 28, 2015 1:25:55 AM com.dropbox.djinni.NativeLibLoader loadLibrary
     [java] INFO: Loaded djinni/djinni/test-suite/java/build/local/lib/libDjinniTestNative.dylib
     [java] WARNING: JNI local refs: 36, exceeds capacity: 35
     [java] 	at org.junit.runner.JUnitCore.<init>(JUnitCore.java:29)
     [java] 	at org.junit.runner.JUnitCore.runMainAndExit(JUnitCore.java:47)
     [java] 	at org.junit.runner.JUnitCore.main(JUnitCore.java:40)
     [java] 	at com.dropbox.djinni.test.AllTests.main(AllTests.java:29)
     [java] JUnit version 4.11
```

**Comment by tonygrue** [comment]
Some Googling indicates some warnings were added at last year in JDK8.

https://bugs.openjdk.java.net/browse/JDK-8046668

related discussions:
https://community.oracle.com/message/13290783
https://bugs.openjdk.java.net/browse/JDK-8046668

Looks like there isn't necessarily wrong when seeing this warning (unlike the missed exception checking ones) though to tone down the annoying we may want to preemptively release local refs, use PushLocalFrame/PopLocalFrame, or use EnsureLocalCapacity to provide more space up front (this last one seems brittle).  Or if we can find some way, suppressing these particular highly conservative local ref warnings without supporting the other warnings (like checking for exceptions) might be reasonable.

**Comment by j4cbo** [comment]
Sounds like an off-by-one on our part, since djinni generally does try to be careful with LocalRef accounting (E.g. by PushLocalFrame-ing sufficient capacity)

**Comment by artwyman** [comment]
I don't think it's an off-by-one.  I think it's a case where we're not manually freeing JNI local refs, but depending on them being freed on return to Java, which is a legitimate thing to do when the usage local and bounded.  The warnings I see are "35 exceeds 34" followed by "36 exceeds 35" which I interpret as simply printing each time we go over the previously-allocated capacity by one, presuming 34 is the default capacity, plus whatever extra we might've allocated.

Regardless, this seems to be happening during init, so it's the class initializers which must be behind it.  Since those are implemented in a helper class which doesn't know where it might be called, it's probably not actually legit for them to depend on free-on-return behavior, so this might be an actual leak of refs they're creating but not cleaning up.  I'll investigate.

**Comment by artwyman** [comment]
Yep, found it.
Fixed in commit @53fd6bc70f546e46986b423a3a3d5ca137db0733.
